### ISSUE TITLE
Add extended CM imports to the Game module

### ISSUE THREAD

**Issue by dnk777** [Qfusion/qfusion#402]
I'm currently working on another improvement/glitch fix addition to the bots code. The addition converts old "dangers detector" code (actually the first part of the bot code I have started working on back in the days) into a sophisticated "perception manager", that not only detects entities a bot should dodge, but inspects the full environment, listens to in-game events (an AI hook is added to `G_AddEvent`) and tries to draw a logical conclusion about the enemy origin (like, when all teammates are in bot's fov and a shot event occurs somewhere behind, an enemy is there). Tests for most events/entities involve tracing. Since these traces can span the entire level and thus be very expensive, an efficient cut off is a must. Along with pure algorithmical solutions (cutting of by an event/entity-specific logic), a PVS test is used as a fallback. 

I have noticed that there is a repetitive pattern in these kinds of tests.
```
for(...) {
	if( trap_inPVS( v0, v[i] ) {
            ...
}
```

where v0 remains the same during the loop execution. Looking at the `CM_inPVS` code a redundant tree traversal can be spotted

```
bool CM_InPVS( cmodel_state_t *cms, const vec3_t p1, const vec3_t p2 ) {
	int leafnum;
	int cluster;
	int area1, area2;
	uint8_t *mask;

	leafnum = CM_PointLeafnum( cms, p1 );
	^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

It just feels... wrong, wasting cycles for no reason. Moreover the second points leafnums can be cached too for tests over the same set of destination points from different orgins (like tests for entities/events from different bots POVs). It is easy to fix by adding and exporting another version of `CM_inPVS()` that takes leaf nums instead of vectors into the Game module, along with `CM_PointLeafNum()` (and `CM_BoxLeafNums()` for some test). There is a trouble, these exports have C linkage and does not allow overloads. 

Can you suggest a proper signature for it? I lean to 
```
trap_CM_LeafsInPVS( int leafnum1, int leafnum2 );
```

IIRC the same pattern can be met in scripts too, I might add these calls to script exports as well.

**Comment by viciious** [comment]
I'd rather cache 10 results of CM_PointLeafnum than export new superfluous CM_ methods

**Comment by dnk777** [comment]
Do you mean caching at `qcommon` level? Assuming a vector is the only kind of input for PVS calls, either a rather complex hash map code has to be added to the collision code, or a sequential search has to be done killing the performance won.

What the proposed changes are intended to be used for. Assume a simple lookup table (an array) consisting of entries like 
```
{ 
    int64_t computedAt;
    int leafNum; 
}
```
where an `i`-th entity maps to an `i`-th entry. The `computedAt` field is here to support lazy computation of the leaf num without need to explicitly clean the table each frame.

Let us assume there are `N` entities and PVS test is performed from each entity to each entity, so `( N * N ) / 2` PVS calls have to be done (we assume that the PVS relation is symmetrical). In getting and caching of a leaf num is not available, there would be `N * N` BSP traversals. If there is a way to get and cache a leaf num, the tree traversals count is cut down to `N` times. Thats why I ask for allowing to modify imports.

**Comment by dnk777** [comment]
I have meant caching few recent results is not enough. I want the leaf num to be computed only once per frame for each entity.

**Comment by viciious** [comment]
I dunno, this stuff should all be low priority compared to fixing the
inability to freely change MAX_EDICTS and various related hacks that you
have introduced

27 авг. 2017 г. 7:49 PM пользователь "dnk777" <notifications@github.com>
написал:

**Comment by dnk777** [comment]
I have discovered that most of the proposed calls (like `BoxLeafNums()`, `AreasConnected`) are already exported and used for linking an entity to collision/vis grid, so there is no even need to maintain a separate mapping from entities to cluster/area/leaf nums.

> inability to freely change MAX_EDICTS

It has been fixed in a patch that is in development. I was way too pessimistic with it trying to avoid a cost of saving/restoring movement states in bot movement prediction (if a structure is small enough memcpy calls might be avoided), but it is really insignificant compared to collision, so I have reverted most nasty packing attempts.

The patch contains many tiny improvements and glitch fixes for the bot code (it is not really large), so finally I can concentrate of testing gametype compatibility and getting a sane bot behavior in various gametypes, but it is kept in my local private branch. No commits addition is planned but I want to scrutinize the changed stuff and rewrite/squash some commits, so I need some extra time to finish it.
